### ISSUE TITLE
Ideas for the far future

### ISSUE THREAD

**Issue by lindeloev** [lindeloev/mcp#23]
Here are some ideas which could use some discussion and careful consideration.

# 1. Variance change
Detecting changes in `sigma`. One solution is to do `var(y) ~ 1 ~ 1` meaning "change in intercept of var(y)" with the default being `mean(y) ~1 ~ 1 + x` ("mean response"). So for segment 2, `1 ~~ x` is simply shorthand for `y ~ 1 ~ x` which again is shorthand for `mean(y) ~ 1 ~ x`. Could then be altered to `var(y) ~ 1 ~ x`. For example

    segments = list(
        y ~ 1 + x,  # segment 1
        1 ~ 1 + x, # seg2: change in slope and intercept in mean
        var(y) ~ 1 ~ 1,  # seg2: intercept-change in variance
        1 ~ 0  # Plateau in mean, segment 3, keeps "new" variance from seg2.
    )

I'm not too happy about this. Will need to look how other packages approach this.

# 2. Multivariate
Multiple response variables predicted from single change points. Something like

    segments = list(
        c(y1, y2) ~ 1 + x,  # segment 1
        1 ~ 0  # seg2: joined intercept
    )

This could be merged with 1. to specify a change in one response but not the other:

    segments = list(
        c(y1, y2) ~ 1 + x,  # segment 1
        1 ~ 0  # seg2: joined intercept on y1 and y2, segment 2
        y1 ~ x  # seg3: slope change in y1 but not y2
    )

# 3. Interaction effects with categorical variables
X and Y have to be (modeled as) continuous. But there can be interactions with categorical variables that just correspond to distinct X-Y regression lines. A change point would be the X at which the interaction effect changes. So:

    segments = list(
        y ~ 1 + x*group,  # seg1: int_1 + x2 + int_group2 + x2 * int_group2
        1 ~ 1 + x*group  # seg2: change in this "relationship"
    )

Now that I think about it - could any number of predictors work, as long as the changes are only modeled on one continuous variable, `x`?

**Comment by josswright** [comment]
I've just come across this package whilst coding up some changepoint models directly. The specific application I've been working in is word/letter frequencies for stylometry. I'm interested to know if categorical/multinomial models could fit into this framework?

(This package looks amazing, by the way!)

**Comment by lindeloev** [comment]
@josswright Yes, mathematically and in JAGS it's straightforward to extend mcp to include multinomial models, as long as there is a continuous predictor. Just need mcp to count the number of categories in the dependent variable (N), to set up a logistic model for each of them (`dbern`), and to set up a proper parameterization of the priors on the regression coefficients (probably a dirichlet). The change point will then define a location on `x` where the N-1 regression coefficients change.

I hope to get time to do this after adding survival analysis, but any pull requests are welcome :-)

**Comment by josswright** [comment]
@lindeloev Fantastic! I'm more familiar with Stan than JAGS, but if I find the time then I'll definitely take a look!

**Comment by jpzhangvincent** [comment]
I'm working on abrupt time series forecasting problem and interested in change point/anomaly detection integrated with forecasting algorithms. This package is very promising. Just curious, are there particular reasons to choose JAGS over Stan as the underlying Bayesian software? IMO, stan gains more popularity and extensibility in the community, like "prophet", "tidybayes" and "shinystan". I'm wondering whether it would be considered to include an option to choose "stan" as a backend so we can get a stan model output to integrate with other "stan" based packages.

**Comment by lindeloev** [comment]
@jpzhangvincent I chose JAGS because:

1. Ease of installation. I have had many students struggle installing RTools. I myself recently tried installing RTools4 without success.
2. Total speed. Many problems are solved in ~2-5 seconds in JAGS. But will take 30+ seconds because of compilation with stan. I did see that stan was working on some improvements that will bring down compilation time down considerably, though (?),
3. I'm more familiar with JAGS :-)

If `mcp` is beginning to get more widely adopted, I would definitely add stan support for the reasons your state! I anticipated this early on so it would not require a major re-write. Only a few hundred of the ~2000 lines of code in `mcp` are JAGS-specific and the others should be quite sampler-agnostic.

I'd be happy to hear your thoughts here!

**Comment by jpzhangvincent** [comment]
> @jpzhangvincent I chose JAGS because:
> 
> 1. Ease of installation. I have had many students struggle installing RTools. I myself recently tried installing RTools4 without success.
> 2. Total speed. Many problems are solved in ~2-5 seconds in JAGS. But will take 30+ seconds because of compilation with stan. I did see that stan was working on some improvements that will bring down compilation time down considerably, though (?),
> 3. I'm more familiar with JAGS :-)
> 
> If `mcp` is beginning to get more widely adopted, I would definitely add stan support for the reasons your state! I anticipated this early on so it would not require a major re-write. Only a few hundred of the ~2000 lines of code in `mcp` are JAGS-specific and the others should be quite sampler-agnostic.
> 
> I'd be happy to hear your thoughts here!

Thanks for your response! I'd like to deep dive into the codebase and see anything I can contribute to. Can you give me some pointers where I can better understand how the model is specified in JAGS?

**Comment by lindeloev** [comment]
@jpzhangvincent Yes, definitely! 

`fit = mcp(...)` returns an `mcpfit` object and you can see the JAGS code using `cat(fit)`. 

This code is mostly generated in `R/get_jagscode.R` and `R/get_formula.R` using the abstract representation (the "Segment table") which is also saved to `fit$.other$ST` - one row per segment. ST is not a fixed API but will be in flux as new features are added and old features are refined.

Adding support for `stan` would be the business of making a `get_stancode.R` and add a function `get_formula_str_stan` to the `get_formula.R` file.
